aioxmpp/bookmarks/service.py

```diff
@@ -57,29 +92,256 @@ class BookmarkClient(service.Service):
+                old_entry = old.pop()
+                new_entry = new.pop()
+                if old_entry == new_entry:
+                    # the bookmark is unchanged, do not emit an event
+                    pass
+                else:
+                    self.logger.debug("changed %s -> %s", old_entry, new_entry)
+                    self.on_bookmark_changed(old_entry, new_entry)
+            elif len(new) == 0:
+                for removed in old:
+                    self.logger.debug("removed %s", removed)
+                    self.on_bookmark_removed(removed)
+            elif len(old) == 0:
+                for added in new:
+                    self.logger.debug("added %s", added)
+                    self.on_bookmark_added(added)
+            else:
+                old, new = subdivide(0, old, new)
+
+                assert len(old) == 0 or len(new) == 0
+
+                for removed in old:
+                    self.logger.debug("removed %s", removed)
+                    self.on_bookmark_removed(removed)
+
+                for added in new:
+                    self.logger.debug("added %s", added)
+                    self.on_bookmark_added(added)
+
+        self._bookmark_cache = new_bookmarks
+
+    @asyncio.coroutine
+    def get_bookmarks(self):
+        """
+        Get the stored bookmarks from the server. Causes signals to be
+        fired to reflect the changes.
+
+        :returns: a list of bookmarks
+        """
+        bookmarks = yield from self._get_bookmarks()
+        self._diff_emit_update(bookmarks)
+        return bookmarks
+
+    @asyncio.coroutine
+    def set_bookmarks(self, bookmarks):
+        """
+        Store the sequence of bookmarks `bookmarks`.
+
+        Causes signals to be fired to reflect the changes.
+
+        .. note:: This should normally not be used unless you have to
+                  modify or add many bookmarks at once. To change only
+                  a single or a few bookmarks use the methods
+                  :meth:`add_bookmark`, :meth:`remove_bookmark` and
+                  :meth:`update_bookmark`.
+        """
+        yield from self._set_bookmarks(bookmarks)
+        self._diff_emit_update(bookmarks)
+        return bookmarks
+
+    @asyncio.coroutine
+    def sync(self):
+        """
+        Sync the bookmarks between the local representation and the
+        server.
+
+        This must be called periodically to assure that the signals
+        are fired.
+        """
+        yield from self.get_bookmarks()
+
+    @asyncio.coroutine
+    def add_bookmark(self, new_bookmark):
+        """
+        Add a bookmark.
+
+        Already existant bookmarks are not added twice.
+        """
+        bookmarks = yield from self._get_bookmarks()
+        for bookmark in bookmarks:
+            if bookmark == new_bookmark:
+                break
+        else:
+            bookmarks.append(new_bookmark)
+        yield from self._set_bookmarks(bookmarks)
```

Comment:
This is not complete yet, I think. We should be checking that the remote bookmarks actually contain the bookmark we wanted to add, and if not, add it again (using the freshly retrieved set of bookmarks as a baseline). With a configurable limit of re-tries (3 maybe?). (We will need this re-tried write for pubsub, too.) Only after all that is done, the _diff_emit_update should be called.

If you don’t want to put this in this PR, I’m fine with merging this and handling it in a future update (but it is part of #123, so this PR would not close #123).